pkg/fqdn/name_manager.go

```diff
@@ -257,3 +257,28 @@ func (n *NameManager) updateIPsForName(lookupTime time.Time, dnsName string, new
 	// from before also expired.
 	return (len(cacheIPs) == 0 && len(sortedNewIPs) == 0) || !ip.SortedIPListsAreEqual(sortedNewIPs, cacheIPs)
 }
+
+var ipcacheResource = ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindDaemon, "", "fqdn-name-manager")
+
+// upsertMetadata adds an entry in the ipcache metadata layer for the set of IPs
+func (n *NameManager) upsertMetadata(ips []netip.Addr) {
+	prefixes := make([]netip.Prefix, 0, len(ips))
+	for _, ip := range ips {
+		prefixes = append(prefixes, netip.PrefixFrom(ip, ip.BitLen()))
+	}
+	n.config.IPCache.UpsertPrefixes(prefixes, source.Generated, ipcacheResource)
```

Comment:
Shouldn't this wait on the channel now that `UpsertPrefixes` returns one?

Reply:
That’s the plan. However, we can short-cut in the case where all the IPs already have identities. I need to find the right place to do this.

Reply:
Also, we can drop the NameManager lock before waiting on the channel. This just needs a bit of thought.

Reply:
OK, I've figured out the concurrency story reasonably well. I'll describe it more in the PR description.